README.md
[![](https://img.shields.io/npm/dm/@monterosa/react-native-parallax-scroll.svg?style=flat-square)](https://www.npmjs.com/package/@monterosa/react-native-parallax-scroll)

# react-native-parallax-scroll

A `ScrollView`-like component that:

* Has a parallax background
* Has a parallax foreground
* Has a fixed or sticky header
* Can be nested within other views(FlatList, SectionList)
* Works on iOS and Android

## Installation

```
$ npm install @monterosa/react-native-parallax-scroll --save
```

## Demo

YouTube - https://www.youtube.com/watch?v=zM4koduK32Y

| ![](./demo/ios-demo-4.gif) | ![](./demo/ios-demo-1.gif) | ![](./demo/ios-demo-2.gif) |
![](./demo/ios-demo-3.gif) | ![](./demo/demo-1.gif) | ![](./demo/demo-2.gif) |

## Basic Usage

```js
import ParallaxScroll from '@monterosa/react-native-parallax-scroll';

// Inside of a component's render() method:
render() {
  return (
     <ParallaxScroll
      renderHeader={({ animatedValue }) => <Header animatedValue={animatedValue} />}
      headerHeight={50}
      isHeaderFixed={false}
      parallaxHeight={250}
      renderParallaxBackground={({ animatedValue }) => <Background animatedValue={animatedValue} />}
      renderParallaxForeground={({ animatedValue }) => <Foreground animatedValue={animatedValue} />}
      parallaxBackgroundScrollSpeed={5}
      parallaxForegroundScrollSpeed={2.5}
    >
      <Welcome />
    </ParallaxScroll>
  );
}
```

## Examples

Please clone the repo and run `npm run storybook` or `yarn storybook` to show examples of usages.

## Usage (API)

All of the properties of `ScrollView` are supported. Please refer to the
[`ScrollView` documentation](https://facebook.github.io/react-native/docs/scrollview.html) for more detail.

The `ParallaxScroll` component adds a few additional properties, as described below.

| Property                        | Type                                       | Defaut                                    | Description                                                                                                                                                                                                                                                                                                                                                                                                        |
| ------------------------------- | ------------------------------------------ | ----------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `style`                         | `object`                                   | `{}`                                      | Component's styles                                                                                                                                                                                                                                                                                                                                                                                                 |
| `width`                         | `number`                                   | `Dimensions.get('window').width`          | Component's width.                                                                                                                                                                                                                                                                                                                                                                                                 |
| `height`                        | `number`                                   | `Dimensions.get('window').height`         | Component's height.                                                                                                                                                                                                                                                                                                                                                                                                |
| `innerRef`                      | `func`                                     | `null`                                    | To get a reference to the scrollable component.                                                                                                                                                                                                                                                                                                                                                                    |
| `scrollStyle`                   | `object`                                   | `{}`                                      | These styles will be applied to the scroll view.                                                                                                                                                                                                                                                                                                                                                                   |
| `headerHeight`                  | `number`                                   | `45`                                      | This is the height of sticky(fixed) header.                                                                                                                                                                                                                                                                                                                                                                        |
| `renderHeader`                  | `({ width, height, animatedValue }) => {}` | `null`                                    | This renders an optional sticky(fixed) header that will be visible to the top of the view.                                                                                                                                                                                                                                                                                                                         |
| `onHeaderFixed`                 | `func`                                     | `null`                                    | A callback function that is invoked when the header will attach to the top.                                                                                                                                                                                                                                                                                                                                        |
| `isHeaderFixed`                 | `bool`                                     | `false`                                   | Is header fixed to top(not sticky)?                                                                                                                                                                                                                                                                                                                                                                                |
| `parallaxHeight`                | `number`                                   | `Dimensions.get('window').width * 9 / 16` | This is the height of parallax.                                                                                                                                                                                                                                                                                                                                                                                    |
| `useNativeDriver`               | `bool`                                     | `false`                                   | Enable [Native driver](https://facebook.github.io/react-native/blog/2017/02/14/using-native-driver-for-animated.html) for animated. NOTE: Works only with `Animated.ScrollView` component.                                                                                                                                                                                                                         |
| `scrollableComponent`           | `class`                                    | `Animated.ScrollView`                     | This is a class of scrollable component.                                                                                                                                                                                                                                                                                                                                                                           |
| `isBackgroundScalable`          | `bool`                                     | `true`                                    | Is background scalable on iOS?                                                                                                                                                                                                                                                                                                                                                                                     |
| `headerBackgroundColor`         | `string`                                   | `rgba(0, 0, 0, 0)`                        | The color of the unsticked(unfixed) header background. Can be empty `''` string. NOTE: Dosen't work with useNativeDriver.                                                                                                                                                                                                                                                                                          |
| `contentContainerStyle`         | `object`                                   | `{}`                                      | These styles will be applied to the scroll view content container which wraps all of the child views.                                                                                                                                                                                                                                                                                                              |
| `headerFixedTransformY`         | `number`                                   | `0`                                       | This number indicating how much the fixed header should move upwards during the scroll. Used as the hack to change fixed header height during scroll.                                                                                                                                                                                                                                                              |
| `onChangeHeaderVisibility`      | `func`                                     | `null`                                    | A callback function that is invoked when the parallax header is hidden or shown (as the user is scrolling). Function is called with a `boolean` value to indicate whether header is visible or not.                                                                                                                                                                                                                |
| `renderParallaxBackground`      | `({ width, height, animatedValue }) => {}` | `null`                                    | This renders the background of the parallax.                                                                                                                                                                                                                                                                                                                                                                       |
| `renderBackgroundPlaceholder`   | `({ height, animatedValue }) => {}`        | `null`                                    | By default we assume that you want to show the foreground and background in its own space. We prepend an empty view with the height of `parallaxHeight`. You can override this behaviour to layer things in between foreground and background. You might want to avoid foreground in such a scenario, see Issue [#23](https://github.com/monterosalondon/react-native-parallax-scroll/issues/23) for more details. |
| `renderParallaxForeground`      | `({ width, height, animatedValue }) => {}` | `null`                                    | This renders the foreground of the parallax.                                                                                                                                                                                                                                                                                                                                                                       |
| `fadeOutParallaxBackground`     | `bool`                                     | `false`                                   | If `true`, the background will fade out as the user scrolls up.                                                                                                                                                                                                                                                                                                                                                    |
| `fadeOutParallaxForeground`     | `bool`                                     | `false`                                   | If `true`, the foreground will fade out as the user scrolls up.                                                                                                                                                                                                                                                                                                                                                    |
| `headerFixedBackgroundColor`    | `string`                                   | `rgba(0, 0, 0, 1)`                        | The color of the sticked(fixed) header background.                                                                                                                                                                                                                                                                                                                                                                 |
| `backgroundScale`               | `number`                                   | `3`                                       | The speed factor that the background scales. Respects `backgroundScaleOrigin`                                                                                                                                                                                                                                                                                                                                      |
| `backgroundScaleOrigin`         | `center` or `top`                          | `center`                                  | The point of which the background should scroll from. Use `top` with `backgroundScale={2}` in order to achieve iOS native parallax scroll behavior. Top will keep the background locked to the top of the scrollview, while scaling the image such that it stretches the background in the downwards direction.                                                                                                    |
| `parallaxBackgroundScrollSpeed` | `number`                                   | `5`                                       | The speed factor that the background moves at relative to the scroll content.                                                                                                                                                                                                                                                                                                                                      |
| `parallaxForegroundScrollSpeed` | `number`                                   | `5`                                       | The speed factor that the foreground moves at relative to the scroll content.                                                                                                                                                                                                                                                                                                                                      |

## Latest changelog

All changes [here](./CHANGELOG.md)

### 1.8.0 - 2018-04-14

#### Added

* `renderBackgroundPlaceholder` prop for overriding default behaviour to layer things in between foreground and background (#23)

### 1.7.0 - 2018-04-12

#### Added

* `backgroundScale` and `backgroundScaleOrigin` props for support scaling from the top and make scale factor configurable (#22)

### 1.6.0 - 2018-04-05

#### Fixed

* Based on the Component, instead of PureComponent

## Contributing

I welcome contributions! Please open an issue if you have any feature ideas
or find any bugs. I also accept pull requests with open arms. I will
go over the issues when I have time. :)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

### 1.8.0 - 2018-04-14

#### Added

* `renderBackgroundPlaceholder` prop for overriding default behaviour to layer things in between foreground and background (#23)

### 1.7.0 - 2018-04-12

#### Added

* `backgroundScale` and `backgroundScaleOrigin` props for support scaling from the top and make scale factor configurable (#22)

### 1.6.0 - 2018-04-05

#### Fixed

* Based on the Component, instead of PureComponent

### 1.5.4 - 2018-03-03

#### Fixed

* header translateY when header is fixed and `headerFixedTransformY` is too big

### 1.5.3 - 2018-02-14

#### Fixed

* issues #17, #18 - fatal error when `scrollableComponent` prop is missing

### 1.5.2 - 2018-02-11

#### Fixed

* issue #15 - scroll to top if scrollableComponent is not Animatable

### 1.5.1 - 2017-12-01

#### Fixed

* onHeaderFixed - Header interpolation when parallxHeight === headerHeight

### 1.5.0 - 2017-11-06

#### Added

* onHeaderFixed - A callback function that is invoked when the header will attach to the top
* headerFixedTransformY - This number indicating how much the fixed header should move upwards during the scroll. Used as the hack to change fixed header height during scroll


src/index.js
/* eslint-disable import/no-extraneous-dependencies */
import React, { Component } from 'react';
import { View, Animated, Dimensions, ListViewDataSource } from 'react-native';
import PropTypes from 'prop-types';
/* eslint-enable import/no-extraneous-dependencies */

const window = Dimensions.get('window');

const KEY = '__PARALLAX_SCROLL__';
const RATIO = 9 / 16;

export default class ParallaxScroll extends Component {
  static propTypes = {
    data: PropTypes.oneOfType([PropTypes.array]),
    style: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
    width: PropTypes.number,
    height: PropTypes.number,
    innerRef: PropTypes.func,
    sections: PropTypes.oneOfType([PropTypes.array]),
    onScroll: PropTypes.func,
    children: PropTypes.oneOfType([PropTypes.array, PropTypes.element]),
    renderRow: PropTypes.func,
    renderItem: PropTypes.func,
    dataSource: PropTypes.instanceOf(ListViewDataSource),
    scrollStyle: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
    headerHeight: PropTypes.number,
    renderHeader: PropTypes.func,
    isHeaderFixed: PropTypes.bool,
    onHeaderFixed: PropTypes.func,
    parallaxHeight: PropTypes.number,
    useNativeDriver: PropTypes.bool,
    backgroundScale: PropTypes.number,
    scrollableComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    isBackgroundScalable: PropTypes.bool,
    backgroundScaleOrigin: PropTypes.oneOf(['top', 'center']),
    headerFixedTransformY: PropTypes.number,
    headerBackgroundColor: PropTypes.string,
    contentContainerStyle: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.number,
      PropTypes.object
    ]),
    onChangeHeaderVisibility: PropTypes.func,
    renderParallaxBackground: PropTypes.func,
    renderParallaxForeground: PropTypes.func,
    fadeOutParallaxForeground: PropTypes.bool,
    fadeOutParallaxBackground: PropTypes.bool,
    headerFixedBackgroundColor: PropTypes.string,
    renderBackgroundPlaceholder: PropTypes.func,
    parallaxBackgroundScrollSpeed: PropTypes.number,
    parallaxForegroundScrollSpeed: PropTypes.number
  };

  static defaultProps = {
    data: null,
    style: {},
    width: window.width,
    height: window.height,
    innerRef: null,
    sections: null,
    children: null,
    onScroll: null,
    renderRow: null,
    dataSource: null,
    renderItem: null,
    scrollStyle: {},
    headerHeight: 45,
    renderHeader: null,
    isHeaderFixed: false,
    onHeaderFixed: () => {},
    parallaxHeight: window.width * RATIO,
    backgroundScale: 3,
    useNativeDriver: false,
    scrollableComponent: Animated.ScrollView,
    isBackgroundScalable: true,
    backgroundScaleOrigin: 'center',
    headerFixedTransformY: 0,
    headerBackgroundColor: 'rgba(0, 0, 0, 0)',
    contentContainerStyle: {},
    onChangeHeaderVisibility: () => {},
    renderParallaxBackground: null,
    renderParallaxForeground: null,
    fadeOutParallaxForeground: false,
    fadeOutParallaxBackground: false,
    headerFixedBackgroundColor: 'rgba(0, 0, 0, 1)',
    renderBackgroundPlaceholder: null,
    parallaxBackgroundScrollSpeed: 5,
    parallaxForegroundScrollSpeed: 5
  };

  scrollY = new Animated.Value(0);
  isHeaderFixed = false;
  isHeaderVisibible = true;

  constructor(props) {
    super(props);

    this.scrollableComponent = props.scrollableComponent;

    if (Animated.ScrollView !== props.scrollableComponent) {
      this.scrollableComponent = Animated.createAnimatedComponent(props.scrollableComponent);
    }

    this._onAnimatedScroll = Animated.event(
      [{ nativeEvent: { contentOffset: { y: this.scrollY } } }],
      {
        listener: this._onScroll,
        useNativeDriver: props.useNativeDriver
      }
    );
  }

  // eslint-disable-next-line
  _getFlatData(data) {
    return [{ key: KEY }, ...data];
  }

  // eslint-disable-next-line
  _getDataSource(dataSource) {
    return dataSource.cloneWithRowsAndSections({
      [KEY]: [''],
      ...dataSource._dataBlob
    });
  }

  _getSectionData(sections) {
    if (this.props.renderItem) {
      return [{ data: [{ key: KEY }], key: KEY }, ...sections];
    }

    return [
      { data: [{ key: KEY }], key: KEY, renderItem: this._renderBackgroundPlaceholder },
      ...sections
    ];
  }

  _onScroll = e => {
    const contentOffsetY = e.nativeEvent.contentOffset.y;
    const {
      onScroll,
      renderHeader,
      headerHeight,
      isHeaderFixed,
      onHeaderFixed,
      parallaxHeight,
      headerFixedTransformY,
      onChangeHeaderVisibility
    } = this.props;

    const isHeaderFixedAfterScroll =
      contentOffsetY > parallaxHeight - headerHeight + headerFixedTransformY;
    const isHeaderVisibibleAfterScroll = contentOffsetY < parallaxHeight;

    if (onScroll) {
      onScroll(e);
    }

    if (renderHeader && isHeaderFixed && this.isHeaderFixed !== isHeaderFixedAfterScroll) {
      this.isHeaderFixed = isHeaderFixedAfterScroll;
      onHeaderFixed(isHeaderFixedAfterScroll);
    }

    if (renderHeader && !isHeaderFixed && this.isHeaderVisibible !== isHeaderVisibibleAfterScroll) {
      this.isHeaderVisibible = isHeaderVisibibleAfterScroll;
      onChangeHeaderVisibility(isHeaderVisibibleAfterScroll);
    }
  };

  _renderRow = (rowData, sectionID, rowID, highlightRow) => {
    if (sectionID === KEY) {
      return this._renderBackgroundPlaceholder();
    }

    return this.props.renderRow(rowData, sectionID, rowID, highlightRow);
  };

  _renderItem = e => {
    if (e.item.key === KEY) {
      return this._renderBackgroundPlaceholder();
    }

    return this.props.renderItem(e);
  };

  _renderBackgroundPlaceholder = () => {
    const { parallaxHeight, renderBackgroundPlaceholder } = this.props;

    if (renderBackgroundPlaceholder) {
      return renderBackgroundPlaceholder({
        animatedValue: this.scrollY,
        height: parallaxHeight
      });
    }

    return <View style={{ height: parallaxHeight }} />;
  };

  _ref = ref => {
    if (typeof this.props.innerRef === 'function' && ref && ref._component) {
      this.props.innerRef(ref._component);
    }
  };

  _renderParallaxBackground() {
    const {
      width,
      parallaxHeight: height,
      isBackgroundScalable,
      renderParallaxBackground,
      fadeOutParallaxBackground,
      parallaxBackgroundScrollSpeed,
      backgroundScale,
      backgroundScaleOrigin
    } = this.props;

    const topOrigin = backgroundScaleOrigin === 'top';

    const translateY = !height
      ? 0
      : this.scrollY.interpolate({
        inputRange: [...(topOrigin ? [-height, 0] : [0]), height],
        outputRange: [
          ...(topOrigin ? [-(height / backgroundScale) + height, 0] : [0]),
          -(height / parallaxBackgroundScrollSpeed)
        ],
        extrapolateLeft: 'extend',
        extrapolateRight: 'extend'
      });

    const scale =
      !isBackgroundScalable || !height
        ? 1
        : this.scrollY.interpolate({
          inputRange: [-height, 0],
          outputRange: [backgroundScale, 1],
          extrapolateLeft: 'extend',
          extrapolateRight: 'clamp'
        });

    const opacity =
      !fadeOutParallaxBackground || !height
        ? 1
        : this.scrollY.interpolate({
          inputRange: [0, height],
          outputRange: [1, 0],
          extrapolate: 'clamp'
        });

    return (
      <Animated.View
        style={{
          position: 'absolute',
          width,
          height,
          opacity,
          transform: [{ translateY }, { scale }]
        }}
        pointerEvents="box-none"
      >
        {renderParallaxBackground({ width, height, animatedValue: this.scrollY })}
      </Animated.View>
    );
  }

  _renderParallaxForeground() {
    const {
      width,
      parallaxHeight: height,
      renderParallaxForeground,
      fadeOutParallaxForeground,
      parallaxForegroundScrollSpeed
    } = this.props;

    /* eslint-disable indent */
    const translateY = !height
      ? 1
      : this.scrollY.interpolate({
          inputRange: [0, height],
          outputRange: [0, -(height / parallaxForegroundScrollSpeed)],
          extrapolateRight: 'extend',
          extrapolateLeft: 'clamp'
        });

    const opacity =
      !fadeOutParallaxForeground || !height
        ? 1
        : this.scrollY.interpolate({
            inputRange: [0, height],
            outputRange: [1, 0],
            extrapolateRight: 'extend',
            extrapolateLeft: 'clamp'
          });
    /* eslint-disable indent */

    const wrapperStyle = {
      position: 'absolute',
      top: 0,
      width,
      height,
      zIndex: 1
    };

    return (
      <View style={wrapperStyle} pointerEvents="box-none">
        <Animated.View
          style={{
            position: 'absolute',
            width,
            height,
            opacity,
            transform: [{ translateY }]
          }}
          pointerEvents="box-none"
        >
          {renderParallaxForeground({ width, height, animatedValue: this.scrollY })}
        </Animated.View>
      </View>
    );
  }

  _renderHeader() {
    const {
      width,
      headerHeight: height,
      renderHeader,
      isHeaderFixed,
      parallaxHeight,
      useNativeDriver,
      headerFixedTransformY,
      headerBackgroundColor,
      headerFixedBackgroundColor
    } = this.props;

    const wrapperStyle = {
      position: 'absolute',
      top: 0,
      width,
      height,
      zIndex: 2
    };
    const style = {
      flex: 1,
      width,
      height
    };

    if (!isHeaderFixed) {
      const parallaxHeightWithoutHeader = parallaxHeight - height;

      style.transform = [
        {
          translateY: this.scrollY.interpolate({
            inputRange: parallaxHeightWithoutHeader
              ? [0, parallaxHeight - height, parallaxHeight]
              : [0, parallaxHeight],
            outputRange: parallaxHeightWithoutHeader ? [0, 0, -height] : [0, -height],
            extrapolate: 'clamp'
          })
        }
      ];
    } else if (headerFixedTransformY) {
      style.transform = [
        {
          translateY: this.scrollY.interpolate({
            inputRange: [0, headerFixedTransformY],
            outputRange: [0, -headerFixedTransformY],
            extrapolate: 'clamp'
          })
        }
      ];
    }

    if (!useNativeDriver && headerBackgroundColor) {
      style.backgroundColor = this.scrollY.interpolate({
        inputRange: [0, parallaxHeight - height],
        outputRange: [headerBackgroundColor, headerFixedBackgroundColor],
        extrapolate: 'clamp'
      });
    } else if (useNativeDriver && headerBackgroundColor) {
      style.backgroundColor = headerBackgroundColor;
    }

    return (
      <View style={wrapperStyle} pointerEvents="box-none">
        <Animated.View style={style} pointerEvents="box-none">
          {renderHeader({ width, height, animatedValue: this.scrollY })}
        </Animated.View>
      </View>
    );
  }

  render() {
    const {
      data,
      style: wrapperStyle,
      width,
      height,
      sections,
      children,
      renderRow,
      renderItem,
      dataSource,
      scrollStyle,
      renderHeader,
      useNativeDriver,
      scrollableComponent,
      contentContainerStyle,
      renderParallaxBackground,
      renderParallaxForeground,
      ...scrollViewProps
    } = this.props;

    const style = [scrollStyle, { width, height }];
    const stickyHeaderIndices = [];
    const ScrollableComponent = this.scrollableComponent;
    const isRenderChildComponents =
      !(data && renderItem) &&
      !((sections && renderItem) || (sections && sections[0].renderItem)) &&
      !(dataSource && renderRow);

    if (isRenderChildComponents && renderParallaxForeground) {
      stickyHeaderIndices.push(stickyHeaderIndices.length);
    }

    return (
      <View style={[wrapperStyle, { width, height }]} onLayout={this._onLayout}>
        {renderParallaxBackground && this._renderParallaxBackground(!!renderHeader)}

        <ScrollableComponent
          {...scrollViewProps}
          ref={this._ref}
          data={data && renderItem && this._getFlatData(data)}
          style={style}
          throttle={16}
          onScroll={this._onAnimatedScroll}
          sections={sections && this._getSectionData(sections)}
          renderRow={dataSource && renderRow && this._renderRow}
          renderItem={renderItem && this._renderItem}
          dataSource={dataSource && renderRow && this._getDataSource(dataSource)}
          scrollEventThrottle={16}
          stickyHeaderIndices={stickyHeaderIndices}
          contentContainerStyle={[contentContainerStyle, { width, minHeight: height }]}
          stickySectionHeadersEnabled
          onMoveShouldSetResponder={() => false}
          onStartShouldSetResponderCapture={() => false}
        >
          {isRenderChildComponents && renderParallaxForeground && this._renderParallaxForeground()}

          {isRenderChildComponents && this._renderBackgroundPlaceholder()}

          {isRenderChildComponents && children}
        </ScrollableComponent>

        {!isRenderChildComponents && renderParallaxForeground && this._renderParallaxForeground()}

        {renderHeader && this._renderHeader()}
      </View>
    );
  }
}
